src/migrate.c

```diff
@@ -0,0 +1,161 @@
+    int argc = 3 + (req->r.migrate_keys.num_keys * 2);
+    char **argv = RedisModule_Calloc(argc, sizeof(char *));
+    size_t *argv_len = RedisModule_Calloc(argc, sizeof(size_t));
+
+    argv[0] = RedisModule_Strdup("RAFT.IMPORT");
+    argv_len[0] = strlen("RAFT.IMPORT");
+    argv[1] = RedisModule_Alloc(32);
+    int n = snprintf(argv[1], 64, "%ld", raft_get_current_term(rr->raft));
+    argv_len[1] = n;
+    argv[2] = RedisModule_Alloc(32);
+
+    for (size_t i = 0; i < req->r.migrate_keys.num_keys; i++) {
+        size_t key_len;
+        const char * key = RedisModule_StringPtrLen(req->r.migrate_keys.keys[i], &key_len);
+        argv[3 + (i*2)] = RedisModule_Strdup(key);
+        argv_len[3+ (i*2)] = key_len;
+
+        enterRedisModuleCall();
+        RedisModuleCallReply * reply = RedisModule_Call(rr->ctx, "DUMP", "c", key);
+        exitRedisModuleCall();
+
+        if (reply && RedisModule_CallReplyType(reply) == REDISMODULE_REPLY_STRING) {
+            size_t str_len;
+            const char * str = RedisModule_CallReplyStringPtr(reply, &str_len);
+            argv[3 + (i*2) + 1] = RedisModule_Alloc(str_len);
+            memcpy(argv[3 + (i*2) + 1], str, str_len);
+            argv_len[3 + (i*2) + 1] = str_len;
+            RedisModule_FreeCallReply(reply);
+        } else {
+            if (reply) {
+                LOG_WARNING("unexpected response type = %d", RedisModule_CallReplyType(reply));
+                RedisModule_FreeCallReply(reply);
+            } else {
+                LOG_WARNING("didn't get a reply!");
+            }
+            ConnAsyncTerminate(conn);
+            RedisModule_ReplyWithError(req->ctx, "ERR see logs");
+            RaftReqFree(req);
+            goto exit;
+        }
+    }
+
+    if (redisAsyncCommandArgv(ConnGetRedisCtx(conn), transferKeysResponse, conn, argc, (const char **) argv, argv_len) != REDIS_OK) {
+        redisAsyncDisconnect(ConnGetRedisCtx(conn));
+        ConnMarkDisconnected(conn);
+    }
+
+exit:
+    for(int i = 0; i < argc; i++) {
+        RedisModule_Free(argv[i]);
+    }
+    RedisModule_Free(argv);
+    RedisModule_Free(argv_len);
+}
+
+
+void MigrateKeys(RedisRaftCtx *rr, RaftReq *req)
+{
+    JoinLinkState *state = RedisModule_Calloc(1, sizeof(*state));
+    state->type = "migrate";
+    state->connect_callback = transferKeys;
+    time(&(state->start));
+    ShardGroup * sg = getShardGroupById(rr, req->r.migrate_keys.shardGroupId);
+    if (sg == NULL) {
+        RedisModule_ReplyWithError(req->ctx, "ERR couldn't resolve shardgroup id");
+        RaftReqFree(req);
+        return;
+    }
+
+    for (unsigned int i = 0; i < sg->nodes_num; i++) {
+        LOG_WARNING("MigrateKeys: adding %s:%d", sg->nodes[i].addr.host, sg->nodes[i].addr.port);
+        NodeAddrListAddElement(&state->addr, &sg->nodes[i].addr);
+    }
+    state->req = req;
+
+    state->conn = ConnCreate(rr, state, joinLinkIdleCallback, joinLinkFreeCallback);
```

Comment:
No need to do right now but probably it is a good idea to keep a connection open until migration is completed. Currently, a timer kicks in every 100 millisecond and establish the connection (this will probably take 1-2 milliseconds on cloud). If we are going to process thousands of migrate commands, then it'll prolong migration time significantly.

Reply:
ok, would have to figure out how to persist it.